spec.yaml

```diff
@@ -218,6 +231,95 @@ definitions:
+          type: object
+          properties:
+            name:
+              type: string
+              description: |
+                A supported resource type (file, folder, calendar, contact, ...).
+                Implementations MUST support `file` at a minimum.
+              example: file
+            shareTypes:
+              type: array
+              description: |
+                The supported recipient share types.
+                Implementations MUST support `user` at a minimum.
+              items:
+                type: string
+                enum: ["user", "group", "federation"]
+              example: ["user"]
+            protocols:
+              type: object
+              description: |
+                The supported protocols to access shares at this endpoint.
+                Implementations MUST support `webdav` at a minimum.
+              anyOf:
+                - properties:
+                    webdav:
+                      type: string
+                      description: |
+                        The top-level WebDAV endpoint URI. In order to access a
+                        remote shared resource, implementations MAY use this
+                        URI as a prefix: e.g. if in a new share request's payload
+                        a `protocol.options.sharedSecret` property is defined
+                        (see singleProtocolLegacy example), it may need to be
+                        appended to this URI to access the resource.
+                      example: https://my-cloud-storage.org/remote.php/dav/ocm
+                - properties:
+                    webapp:
+                      type: string
+                      description: |
+                        The top-level endpoint URI for web apps. This value is
+                        provided for documentation purposes, and it SHALL NOT
+                        be intended as a prefix for share requests.
+                      example: https://my-cloud-storage.org/external/ocm
+                - properties:
+                    datatx:
+                      type: string
+                      description: |
+                        The top-level endpoint URI for data transfers. This value
+                        is provided for documentation purposes, and it SHALL NOT
+                        be intended as a prefix. Furthermore, implementations
+                        are expected to execute the transfer using WebDAV as
+                        the wire protocol.
+                      example: https://my-cloud-storage.org/remote.php/dav/ocm
+      capabilities:
+        type: array
+        description: |
+          The capabilities exposed at this endpoint according to the present
+          specifications. Implementations MUST expose `/shares` at a minimum.
+        items:
+          type: string
+          enum: ["/shares", "/notifications", "/invite-accepted"]
+        example:
+          ["/shares", "/invite-accepted"]
```

Comment:
I like the idea! I see this as a way to advertise additional endpoints.

- Should we use method and resource tuples in place of endpoints? Do we gain anything from that?
- Should mandatory endpoints be included as shown?

Reply:
Good questions! For the second, true that it does not add any information to expose `/shares` if we state that `/shares` must be implemented, so yeah let's drop it. For the first, could you please provide an example to make sure we understand what would be the model?

Reply:
This is what I had in mind originally, but it's rather difficult to express. Mixed arrays (oneOf) is supported in OpenAPI 3.0 and tuple validation (prefixItems) requires OpenAPI 3.1.

```
capabilities:
  type: array
  description: |
    The capabilities exposed at this endpoint according to the present
    specifications. As implementations MUST provide `/shares` to be compliant,
    it is not necessary to expose it as a capability.
  items:
    oneOf:
      - type: string
      - type: array
        prefixItems:
          - type: string
            enum: ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD", "OPTIONS", "TRACE"]
          - type: string
        minItems: 2
        maxItems: 2
  example:
  - /invite-accepted
  - ["POST", "/invite-accepted"]
```

Reply:
I see, yes, again maybe once we migrate the whole spec to OpenAPI v3.0+? But I'm not sure there's any added value to advertise the methods, in the sense that e.g. `/invite-accepted` must be a `POST` action - by the spec itself, and any other optional endpoint must work according to the spec. In other words I don't see a reason to make the method a "capability" of any given endpoint. In that respect, we could keep it simpler with just a list...

Reply:
Agreed